src/pages/instances/actions/InstanceBulkAction.tsx

```diff
@@ -0,0 +1,142 @@
+  instances: LxdInstance[];
+  isLoading: boolean;
+  isDisabled: boolean;
+  confirmationLabel: string;
+  onClick: () => void;
+  confirmationAppearance?: string;
+  confirmationExtra?: ReactNode;
+}
+
+const InstanceBulkAction: FC<Props> = ({
+  action,
+  confirmationAppearance,
+  confirmationExtra,
+  confirmationLabel,
+  icon,
+  instances,
+  isLoading,
+  isDisabled,
+  onClick,
+}) => {
+  const statusLine = (
+    currentState: LxdInstanceStatus,
+    desiredAction: LxdInstanceAction
+  ) => {
+    const count = instances.filter(
+      (instance) => instance.status === currentState
+    ).length;
+
+    const status = statusLabel(currentState) ?? "";
+    const actionRaw = instanceAction(desiredAction, currentState);
+    const action = instanceActionLabel(actionRaw);
+
+    if (count === 0) {
+      return null;
+    }
+
+    if (action === "ignored") {
+      const instance = count === 1 ? "instance that is" : "instances that are";
+      const already = desiredAction !== "restart" ? "already " : "";
+      return (
+        <>
+          - No action for <b>{count}</b> {instance} {already}
+          {currentState.toLowerCase().replace("frozen", "paused")}.
+          <br />
+        </>
+      );
+    }
+
+    const indent = hasDifferentStates ? "- " : "";
+
+    return (
+      <>
+        {indent}
+        <b>{count}</b>
+        {` ${status} ${pluralizeInstance(count)} will be ${action}.`}
+        <br />
+      </>
+    );
+  };
+
+  const selectedStates = new Set(instances.map((item) => item.status));
+  const hasDifferentStates = selectedStates.size > 1;
+  const selectedSummary = hasDifferentStates ? (
+    <>
+      <b>{instances.length}</b> {pluralizeInstance(instances.length)} selected:
+      <br />
+      <br />
+    </>
+  ) : null;
+
+  const hasChangedStates = [...selectedStates].some(
+    (state) => instanceAction(action, state) !== undefined
+  );
+
+  // no action states should be last
+  const getLineOrder = (): LxdInstanceStatus[] => {
+    switch (action) {
+      case "start":
+        return ["Frozen", "Stopped", "Running"];
+      case "restart":
+        return ["Running", "Stopped", "Frozen"];
+      case "freeze":
+        return ["Running", "Stopped", "Frozen"];
+      case "stop":
+        return ["Frozen", "Running", "Stopped"];
+      default:
+        return [];
+    }
+  };
+
+  return (
+    <ConfirmationButton
+      className="u-no-margin--right u-no-margin--bottom bulk-action"
+      confirmationButtonAppearance={confirmationAppearance}
+      confirmButtonLabel={confirmationLabel}
+      confirmationExtra={confirmationExtra}
+      confirmationMessage={
+        <>
+          {selectedSummary}
+          {getLineOrder().map((state) => statusLine(state, action))}
+          <br />
+          Are you sure you want to {confirmationLabel.toLowerCase()}{" "}
+          {instanceActions(instances, action).filter(
+            (item) => item.action !== undefined
+          ).length === 1
```

Comment:
```suggestion
          {instanceActions(instances, action).length === 1
```
In case you update the `instanceActions` function as suggested, `filter`ing here becomes not needed anymore.